### ISSUE TITLE
Compiler requires redundant trait bounds which should be inferrable

### ISSUE THREAD

**Issue by joshlf** [rust-lang/rust#58214]
The following program should compile, [but doesn't](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4bd09b30473286ae681c13e7f7e3b8f6):

```rust
pub trait Foo: Iterator
where
    <Self as Iterator>::Item: Clone,
{
}

#[derive(Clone)]
struct MyWrapper<T>(T);

impl<T: Foo> Foo for MyWrapper<T> {}

impl<T: Foo> Iterator for MyWrapper<T> {
    type Item = T::Item;

    fn next(&mut self) -> Option<T::Item> {
        self.0.next()
    }
}
```

In particular, `rustc` seems unable to infer that the `T: Foo` bound in each of the two impl blocks implies that `<T as Iterator>::Item: Clone`:

```text
error[E0277]: the trait bound `<T as std::iter::Iterator>::Item: std::clone::Clone` is not satisfied
  --> src/lib.rs:10:14
   |
10 | impl<T: Foo> Foo for MyWrapper<T> {}
   |              ^^^ the trait `std::clone::Clone` is not implemented for `<T as std::iter::Iterator>::Item`
   |
   = help: consider adding a `where <T as std::iter::Iterator>::Item: std::clone::Clone` bound

error[E0277]: the trait bound `<T as std::iter::Iterator>::Item: std::clone::Clone` is not satisfied
  --> src/lib.rs:12:1
   |
12 | / impl<T: Foo> Iterator for MyWrapper<T> {
13 | |     type Item = T::Item;
14 | |
15 | |     fn next(&mut self) -> Option<T::Item> {
16 | |         self.0.next()
17 | |     }
18 | | }
   | |_^ the trait `std::clone::Clone` is not implemented for `<T as std::iter::Iterator>::Item`
   |
   = help: consider adding a `where <T as std::iter::Iterator>::Item: std::clone::Clone` bound
note: required by `Foo`
  --> src/lib.rs:1:1
   |
1  | / pub trait Foo: Iterator
2  | | where
3  | |     <Self as Iterator>::Item: Clone,
4  | | {
5  | | }
   | |_^

error[E0277]: the trait bound `<T as std::iter::Iterator>::Item: std::clone::Clone` is not satisfied
  --> src/lib.rs:15:5
   |
15 | /     fn next(&mut self) -> Option<T::Item> {
16 | |         self.0.next()
17 | |     }
   | |_____^ the trait `std::clone::Clone` is not implemented for `<T as std::iter::Iterator>::Item`
   |
   = help: consider adding a `where <T as std::iter::Iterator>::Item: std::clone::Clone` bound
note: required by `Foo`
  --> src/lib.rs:1:1
   |
1  | / pub trait Foo: Iterator
2  | | where
3  | |     <Self as Iterator>::Item: Clone,
4  | | {
5  | | }
   | |_^
```

If, however, we take `rustc`'s advice and add `where <T as Iterator>::Item: Clone` bounds, [it compiles](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38f5f5b7e3d4259b5b6ecaf5ca34e335).

This seems like a bug to me, since the bound we added should be implied by the `T: Foo` bound.

**Comment by jonas-schievink** [comment]
This is a known issue, rustc was never able to propagate bounds like this. Blocked on chalk AFAIK, and also might need an RFC.

**Comment by joshlf** [comment]
Gotcha. Is this tracked in an issue somewhere so I can close as a duplicate of that?

**Comment by jonas-schievink** [comment]
I know we've closed a few dupes of this, but I don't know if there's a canonical issue for this

**Comment by varkor** [comment]
This is [implied bounds](https://github.com/rust-lang/rust/issues/44491).
